### PR TITLE
Fix issue 14854 - @disable this inconsistent between structs and classes

### PR DESCRIPTION
**Pull request by Mai-Lapyst**
This PR adds an deprecation warning for all `@disable` annotations on functions which have a body, so at a later point in time this can be made an error without breaking any existing code surprisingly.

This PR is based on the old PR #9413, but which got closed for unknown reason.

### CODE REVIEW

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
The issue is misguided. @disable has nothing to do with the problem. Structs are not allowed to define default constructors, but classes do not have that limitation. However, both structs and classes may disable default construction. As a consequence, you are allowed to define a default constructor that is disabled for structs, provided that it does not have a body. This is intended behavior. 

Generalizing this to all functions seems like a pointless restriction since you may want to disable different overloads for testing purposes. If we implement this, you would have to comment and uncomment the code of the function body.

In conclusion, I oppose this PR and suggest to close the issue as WONTFIX.

**Comment by Mai-Lapyst** [comment]
> Generalizing this to all functions seems like a pointless restriction since you may want to disable different overloads for testing purposes.

This makes sense; maybe change it to a warning then so one can catch it for release builds or should it rather be implemented in an external tool/linter like d-scanner?

**Comment by RazvanN7** [comment]
We have a general policy against warnings. The compiler should either reject bad code or accept good code. Code smells are the job of linters. So yes, adding a check to dscanner might be preferable.

**Comment by RazvanN7** [comment]
@Mai-Lapyst Thanks for the time and effort put into this. Although the PR was not merged, this has led to the closure of an issue. Congrats!

**Comment by Mai-Lapyst** [comment]
@RazvanN7 thanks, especially for the feedback. learned something regardless of the merge about D and the DMD.
